flake8_trio/visitors/visitor100.py

```diff
@@ -0,0 +1,67 @@
+"""Contains visitor for TRIO100.
+
+A `with trio.fail_after(...):` or `with trio.move_on_after(...):`
+context does not contain any `await` statements.  This makes it pointless, as
+the timeout can only be triggered by a checkpoint.
+Checkpoints on Await, Async For and Async With
+"""
+# if future annotations are imported then shed will reformat away the Union use
+from typing import Any, Union
+
+import libcst as cst
+import libcst.matchers as m
+
+from .flake8triovisitor import Flake8TrioVisitor_cst
+from .helpers import AttributeCall, error_class_cst, with_has_call
+
+
+@error_class_cst
+class Visitor100_libcst(Flake8TrioVisitor_cst):
+    error_codes = {
+        "TRIO100": (
+            "{0}.{1} context contains no checkpoints, remove the context or add"
+            " `await {0}.lowlevel.checkpoint()`."
+        ),
+    }
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.has_checkpoint_stack: list[bool] = []
+        self.node_dict: dict[cst.With, AttributeCall] = {}
+
+    def visit_With(self, node: cst.With) -> None:
+        if m.matches(node, m.With(asynchronous=m.Asynchronous())):
+            self.checkpoint_node(node)
+        if res := with_has_call(
+            node, "fail_after", "fail_at", "move_on_after", "move_on_at", "CancelScope"
+        ):
+            self.node_dict[node] = res
```

Comment:
Here, if we return each match, I think we can `for r in res: self.node_dict[r.node] = r`.  I think that restores reporting each distinct warning if there are multiple calls in a single `with`-statement?

Reply:
leave_with would also have to be changed to loop over it's items (and `leave_WithItem` is not an option since we need to parse it after the body has been parsed). Changing the dict to contain a sequence is probably the way to go.
I went with the lazy option cause I didn't think multiple withitems was used much ever, but I'll fix it in a yiffie~